### PR TITLE
Make Elasticsearch boosting compatible with 'fields' argument

### LINKED ISSUE #11208: Search no longer respects `fields` parameter
### Issue Summary

As shown [here](https://docs.wagtail.org/en/stable/topics/search/searching.html#specifying-the-fields-to-search), the `search` function lets the user specify which fields to search on. 

If any `boost` parameter is specified in `index.SearchField`, all fields are searched even if you specify a limited subset. This seems to be a regression after [this PR](https://github.com/wagtail/wagtail/pull/10653). The query compiler adds a [combined field](https://www.elastic.co/guide/en/elasticsearch/reference/7.13/copy-to.html) with a name similar to `_all_text_boost_2` to each query, even if the `fields` parameter was used. This boosted field contains data from every field, so in effect all fields are searched.

An example of what the compiled query looks like exists in the tests [here](https://github.com/wagtail/wagtail/blob/5cf37f3a9dd586f1a1a86e831bae0a6f1eff2b3b/wagtail/search/tests/test_elasticsearch7_backend.py#L350). I'm not sure that `_all_text_boost_2_0^2.0` and `_all_text_boost_10_0^10.0` should be in this test.

I'm not sure what the intended behavior is. If the goal is only to have `boost` apply when searching all fields, you could only add these boosted fields when `fields` is not specified. Otherwise I think a different way of adding the boost is necessary ([this issue](https://github.com/wagtail/wagtail/issues/5422#issuecomment-645435567)).

I found this issue using the ElasicSearch7 backed. I'm not sure if it applies to database backends. I did not have this issue after removing all `boost` parameters from the models.

### Steps to Reproduce

1. Start a new project and setup ElasicSearch as the search backend. Create a model `Model` with 2 searchable fields, `title` and `author`.
2. Create a dummy entry, such as `Model(title="Title", author="Author"). 
3. Index the data
4. Run the following searches. This is the correct behavior.
```python
backend.search("Title", qs)  # 1 result
backend.search("Author", qs)  # 1 result
backend.search("Title", qs, fields=["author"])  # 0 results
backend.search("Author", qs, fields=["title"])  # 0 results
```
5. Add a `boost` parameter to each searchable field in `Model`
6. Re-index the data
7. Run the following code. This behavior is incorrect.
```python
backend.search("Title", qs)  # 1 result
backend.search("Author", qs)  # 1 result
backend.search("Title", qs, fields=["author"])  # 1 result, should be 0
backend.search("Author", qs, fields=["title"])  # 1 result, should be 0
```

### Technical details

-   Python version: 3.10
-   Django version: 4.2.7
-   Wagtail version: 5.2
-   Browser version: N/A

### PR DESCRIPTION
**Pull request by gasman**
Fixes #11208

The `_all_text_boost_N` index fields (which store the content of all fields that have been boosted to a value of N) were always being inserted into the set of fields to search on, even when an explicit `fields` argument was specified on the query.

Fixing this revealed some previously-masked issues with how boosting was applied on single-field queries (`match` rather than `multi_match`) - in particular, the `"some_field^2"` syntax for boosting is only valid within a `fields` list on `multi_match`, not as the dict key of a `match` query where a `"boost"` option needs to be passed instead.

For future reference: Initially I thought it was incorrect that boosted fields were being indexed within `_all_text` as well as `_all_text_boost_N`, since all-fields searches search on both - however, this is needed for [negated queries such as this one](https://github.com/wagtail/wagtail/blob/1a4595b60b7ade45962ba228c1713d4975fe6a1d/wagtail/search/tests/test_backends.py#L922) to work, since if we only index title text within `_all_text_boost_2`, the `_all_text` field for Javascript: The Definitive Guide ends up containing "javascript" (from the tags) but not "definitive" and is thus wrongly included in the results. I suspect there may still be a bug in the converse case, where the boosted content considered on its own satisfies the "X and not Y" condition, but I think we're a lot closer to correct than we were!

### CODE REVIEW

**Comment by scott-8** [comment]
@gasman That's a good point about `_all_text `. If you switched the values of `title` and `tags` in that test you linked, the boosted field would always match `(PlainText("javascript") & ~PlainText("definitive"))`.

One idea here is to use a boolean operator to make sure the query is found in `_all_text`.

Instead of this:
```json
{
    "bool": {
        "filter": {"match": {"content_type": "searchtests.Book"}},
        "must": {
            "multi_match": {
                "fields": [
                    "_all_text",
                    "_all_text_boost_2_0^2.0",
                    "_all_text_boost_3_0^3.0",
                    "_all_text_boost_4_0^4.0",
                ],
                "query": "Hello",
            },
        },
    }
}
```
It could generate something like this:
```json
{
    "bool": {
        "filter": {"match": {"content_type": "searchtests.Book"}},
        "must": {
            "match": {
                "_all_text": "Hello",
            },
            "multi_match": {
                "fields": [
                    "_all_text_boost_2_0^2.0",
                    "_all_text_boost_3_0^3.0",
                    "_all_text_boost_4_0^4.0",
                ],
                "query": "Hello",
            },
        },
    }
}
```

File: wagtail/search/backends/elasticsearch7.py [hunk 1]
```diff
@@ -587,12 +587,13 @@ def _compile_plaintext_query(self, query, fields, boost=1.0):
         if query.operator != "or":
             match_query["operator"] = query.operator
 
-        if boost != 1.0:
-            match_query["boost"] = boost
-
         if len(fields) == 1:
-            return {"match": {fields[0].field_name_with_boost: match_query}}
+            if boost != 1.0 or fields[0].boost != 1.0:
+                match_query["boost"] = boost * fields[0].boost
```
Comment (laymonage) [inline wagtail/search/backends/elasticsearch7.py:590]: I'm not really sure where this bit of math comes from...?

**Review by laymonage — APPROVED** [review]
There's one bit that I'm not too sure about, but other than that this looks good to me!

File: wagtail/search/backends/elasticsearch7.py (hunk 1 above)
Comment (gasman) [reply wagtail/search/backends/elasticsearch7.py:590]: Boost values can either come from the `SearchField` definition, or [a `Boost` clause within the query](https://docs.wagtail.org/en/stable/topics/search/searching.html#complex-search-queries) - in the latter case, the boost value comes through as the `boost` argument to `_compile_query` and is passed down through recursive calls to `_compile_query`. When boosts exist on both the field and the query, we multiply them to combine their effect - this is consistent with how we handle expressions like `Boost(Boost("query", 2.0), 2.0)` here: https://github.com/wagtail/wagtail/blob/b14bc33cdc93b06257d660256e410dada3beb6ad/wagtail/search/backends/elasticsearch7.py#L677-L678

**Pull request merged**